querqy-core/src/main/java/querqy/rewrite/contrib/ReplaceRewriter.java

```diff
@@ -0,0 +1,111 @@
+package querqy.rewrite.contrib;
+
+import querqy.ComparableCharSequence;
+import querqy.CompoundCharSequence;
+import querqy.LowerCaseCharSequence;
+import querqy.model.AbstractNodeVisitor;
+import querqy.model.Clause;
+import querqy.model.DisjunctionMaxQuery;
+import querqy.model.ExpandedQuery;
+import querqy.model.Node;
+import querqy.model.Query;
+import querqy.model.Term;
+import querqy.rewrite.QueryRewriter;
+import querqy.trie.State;
+import querqy.trie.TrieMap;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static querqy.rewrite.contrib.ReplaceRewriterParser.TOKEN_SEPARATOR;
+
+public class ReplaceRewriter extends AbstractNodeVisitor<Node> implements QueryRewriter {
+
+    private final TrieMap<List<ComparableCharSequence>> replaceRules;
+    private final boolean ignoreCase;
+
+    public ReplaceRewriter(TrieMap<List<ComparableCharSequence>> replaceRules, boolean ignoreCase) {
+        this.replaceRules = replaceRules;
+        this.ignoreCase = ignoreCase;
+    }
+
+    private LinkedList<ComparableCharSequence> querySeq = new LinkedList<>();
+    private LinkedList<ComparableCharSequence> matchSeq = new LinkedList<>();
+    private State<List<ComparableCharSequence>> priorMatch = new State<>(false, null, null);
+    private boolean hasReplacement = false;
+
+    @Override
+    public ExpandedQuery rewrite(ExpandedQuery query) {
+        querySeq.clear();
+        matchSeq.clear();
```

Comment:
I suggest to create these lists only here and not to .clear() them. Rewriters are always stateful and .rewrite() will always be called only once.